### PR TITLE
custom mapper should not always return a 503 when the mapping fails

### LINKED ISSUE #3609: Ccustom mapper should not always return a 503 when the mapping fails
When using a custom http mapper the `HttpBasedMessageMapping` will always return a 503 code when the mapping fails (https://github.com/eclipse-hono/hono/blob/e2fd86459c5d34212464f7579caccb8832cff190/adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java#L253).  This gives the requesting party the assumption that the service is currently not capable of handling the request but maybe it will be able to do so in the future. So when you receive a 503, you might continuously retry doing the request until the server will become ready which might never happen in case you constantly send faulty data.

Preferred behavior would be to return a more suitable error, like just using the status code of the outgoing request when it fails.

### PR DESCRIPTION
**Pull request by JeffreyThijs**
fixes #3609

### CODE REVIEW

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java [hunk 1]
```diff
@@ -279,4 +282,22 @@ private void mapDownstreamMessageRequest(
...
+    }
+
+    private Throwable mapException(final AsyncResult<HttpResponse<Buffer>> httpResponseAsyncResult, final String message) {
+        final Optional<HttpResponse<Buffer>> httpResponse = Optional.ofNullable(httpResponseAsyncResult.result());
+        final int statusCode = httpResponse.map(HttpResponse::statusCode).orElse(HttpURLConnection.HTTP_INTERNAL_ERROR);
+        if (statusCode >= 400 && statusCode < 500) {
+            return new ClientErrorException(statusCode, message);
+        }
+        return new ServerErrorException(statusCode, message);
+    }
```
Comment (sophokles73) [inline adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java:302]: can we reduce the code duplication by e.g. having just one method which accepts an `Optional<String> message`?

File: adapters/mqtt-base/src/test/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMappingTest.java [hunk 2]
```diff
@@ -379,8 +380,8 @@ public void testMappingCommandFailsForWhenPayloadCannotMapped(final VertxTestCon
         messageMapping.mapUpstreamMessage(assertion, command)
             .onComplete(ctx.failing(t -> {
                 ctx.verify(() -> {
-                    assertThat(t).isInstanceOf(ServerErrorException.class);
-                    assertThat((((ServerErrorException) t).getErrorCode())).isEqualTo(HttpURLConnection.HTTP_UNAVAILABLE);
+                    assertThat(t).isInstanceOf(ClientErrorException.class);
+                    assertThat((((ClientErrorException) t).getErrorCode())).isEqualTo(HttpURLConnection.HTTP_FORBIDDEN);
```
Comment (sophokles73) [inline adapters/mqtt-base/src/test/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMappingTest.java:384]: can we also add a test case for when the mapper cannot be invoked at all, i.e. results in an `UNAVAILABLE`?

**Review by sophokles73 — CHANGES_REQUESTED** [review]

File: adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java (hunk 1 above)
Comment (sophokles73) [reply adapters/mqtt-base/src/main/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMapping.java:302]: you might also want to consider using `org.eclipse.hono.client.util.StatusCodeMapper.from()` to create a context specific `ServiceInvocationException` ...

File: adapters/mqtt-base/src/test/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMappingTest.java (hunk 2 above)
Comment (JeffreyThijs) [reply adapters/mqtt-base/src/test/java/org/eclipse/hono/adapter/mqtt/impl/HttpBasedMessageMappingTest.java:384]: i added an additional test.

**Review by sophokles73 — APPROVED** [review]
LGTM

Thanks for contributing :+1:

**Pull request merged**